### DISCUSSION TITLE
mapping '/etc/swanctl/private/privkey.pem' failed: Permission denied, skipped

### DISCUSSION THREAD

**Discussion by CoBool** [strongswan/strongswan#1981] [Q&A]
root@tes:/etc/apparmor.d/local# uname -a
Linux tes 5.15.0-88-generic #98-Ubuntu SMP Mon Oct 2 15:18:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
root@tes:/etc/apparmor.d/local# sudo swanctl --version
strongSwan swanctl 5.9.5
root@tes:/etc/apparmor.d/local# sudo swanctl --load-all
loaded certificate from '/etc/swanctl/x509/cert.pem'
loaded certificate from '/etc/swanctl/x509ca/chain.pem'
mapping '/etc/swanctl/private/privkey.pem' failed: Permission denied, skipped
missing secret in 'ike', ignored
loaded eap secret 'eap'
loaded authority 'letsencrypt'
successfully loaded 1 authorities, 0 unloaded
no pools found, 0 unloaded
loading connection 'ikev2-vpn' failed: invalid value for: proposals, config discarded
loaded 0 of 1 connections, 1 failed to load, 0 unloaded

drwx------   2 root root  4096 Nov  6 21:04 /etc/swanctl/private/

lrwxrwxrwx  1 root root   48 Nov  6 21:04 privkey.pem -> /etc/letsencrypt/live/vpn.tes.com/privkey.pem

-rw------- 1 root root 1765 Aug 27 02:50 cert1.pem
-rw-r--r-- 1 root root 1765 Oct 26 01:50 cert2.pem
-rw-r--r-- 1 root root 3749 Aug 27 02:50 chain1.pem
-rw-r--r-- 1 root root 3749 Oct 26 01:50 chain2.pem
-rw-r--r-- 1 root root 5514 Aug 27 02:50 fullchain1.pem
-rw-r--r-- 1 root root 5514 Oct 26 01:50 fullchain2.pem
-rw------- 1 root root 1704 Aug 27 02:50 privkey1.pem
-rw------- 1 root root 1708 Oct 26 01:50 privkey2.pem

I've been tinkering with permissions and trying different things, but it's not working.

I also created a usr.lib.ipsec.charon file and added it to the 

/etc/letsencrypt/archive/vpn.tes.com/* r, I've added settings, I've done everything, but still the same problem.

Please forgive my poor English.
Thanks for the answer

**Comment by tobiasbrunner** [comment]
> lrwxrwxrwx 1 root root 48 Nov 6 21:04 privkey.pem -> /etc/letsencrypt/live/vpn.tes.com/privkey.pem
> ...
> /etc/letsencrypt/archive/vpn.tes.com/* r, I've added settings, I've done everything, but still the same problem.

Hm, one path is `.../live/...` the other `.../archive/...` (I know the former contains symlinks to the latter, but maybe to AppArmor that doesn't matter, you should check the system log for specific errors).


> loading connection 'ikev2-vpn' failed: invalid value for: proposals, config discarded

You should also fix this.

**Reply by CoBool** [reply]
/etc/apparmor.d/local/usr.lib.ipsec.charon

#include <abstractions/ssl_keys>

I added this one and it worked...
I guess I was too stupid. I'll do some more Googling and try again.

Thanks so much for the answer.
I also fixed the part you told me after consulting with GPT...

**Comment by fancywriter** [comment]
Hello!

Just in case if someone googled this problem (same as I did), the solution is to add the line
```
/etc/letsencrypt/**           r,
```
(or more exact path, I decided for me no harm adding access to all subdirectory).
to the `/etc/apparmor.d/usr.sbin.swanctl` file.

Worked for me.

**Reply by Meinersbur** [reply]
You are my personal hero (worked for Ubuntu 22.04.4 LTS)

The `/etc/apparmor.d/usr.sbin.swanctl` contains this commented config line:
```
  # Allow /etc/swanctl/x509ca/ files to symlink to system-wide ca-certificates
  #include <abstractions/ssl_certs>
```

Then `/etc/apparmor.d/abstractions/ssl_certs` contains:
```
  # certbot
  /etc/letsencrypt/archive/*/cert*.pem r,
  /etc/letsencrypt/archive/*/chain*.pem r,
  /etc/letsencrypt/archive/*/fullchain*.pem r,

  /etc/certbot/archive/*/cert*.pem r,
  /etc/certbot/archive/*/chain*.pem r,
  /etc/certbot/archive/*/fullchain*.pem r,
```
Why only the archive, not the live certificates?

**Reply by tobiasbrunner** [reply]
As I pointed out previously, live certificates are symlinks to the archive, so that's probably why. Also note that you should make changes to files in `/etc/apparmor.d/local/`, so they won't get overwritten by package updates.

**Reply by vadimcoder** [reply]
Thank you!
